core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -223,12 +236,12 @@ abstract class RDD[T: ClassTag](
   }
 
   /** Get the RDD's current storage level, or StorageLevel.NONE if none is set. */
-  def getStorageLevel: StorageLevel = storageLevel
+  def getStorageLevel: StorageLevel = stateLock.synchronized { storageLevel }
```

Comment:
Synchronizing calls on simple getters always raises flags for me. Is this really safe?

For example, in one of the places that calls this (`getOrCompute`): what happens if the storage level changes after this value is returned, but before the call to `blockManager.getOrElseUpdate` actually does its thing?

Reply:
I will be honest, I don't really have a good understanding of how the mutability of storageLevel could be an actual problem.  I do know of an instance where moving a `.cache()` before the creation of a Future which submitted a job fixed a hanging job.  Unfortunately I have very little info about what else was going on in that job.

I have audited the uses of `storageLevel`, like`getOrCompute`, and I think its OK -- there its passing in a local copy, and anyway that whole function is running on the executor where the storageLevel is immutable.

another tricky one is `DAGScheduler.getCacheLocs`, which reads both storage level and `rdd.partitions`.  that gets exposed via `SparkContext.getPreferredLocs`, and used in `CoalescedRDD` and `PartitionAwareUnionRDD`.  but I still can't come up with an explanation for the behavior we see.  I will try to poke some more.

I could also submit a change which just covers `partitions` and `dependencies` for now, since that is clear, though I'd like to understand the problem w/ storageLevel as well.

Reply:
Yeah, what is this racing with? by itself, I'm not sure this forces something to set this before something else needs to read it.